feat(jans-fido2): Attestation and Assertion of Windows Hello platform authenticator

Following are the issues to be fixed:

**Attestation** : the flow proceeds correctly to TPM Processor
1. OID mismatch 
```
java.security.SignatureException: Signature encoding error
    at sun.security.rsa.RSASignature.engineVerify(Unknown Source)
    at java.security.Signature$Delegate.engineVerify(Unknown Source)
    at java.security.Signature.verify(Unknown Source)
    at com.cryptography.VerifyDkimSignature.main(VerifyDkimSignature.java:54)
Caused by: java.io.IOException: ObjectIdentifier mismatch: 1.3.14.3.2.26
    at sun.security.rsa.RSASignature.decodeSignature(Unknown Source)
   ``` 
   Fix:  change https://github.com/JanssenProject/jans/blob/86a7f18f188af9a3eb4cf4f40c7531358b277d42/jans-fido2/server/src/main/java/io/jans/fido2/service/verifier/SignatureVerifier.java#L92 to
   ```
   Signature signatureChecker = Signature.getInstance("SHA1withRSA");
   ```
After this, failures at every step of verification.
Useful References - 
1. https://fy.blackhats.net.au/blog/html/2020/08/24/windows_hello_in_webauthn_rs.html
2. https://github.com/microsoft/InventorySample/tree/master/docs/WindowsHello

--- a/jans-fido2/server/src/main/java/io/jans/fido2/service/processor/attestation/TPMProcessor.java
+++ b/jans-fido2/server/src/main/java/io/jans/fido2/service/processor/attestation/TPMProcessor.java
@@ -19,7 +19,6 @@
 package io.jans.fido2.service.processor.attestation;
 
 import java.io.IOException;
-import java.nio.ByteBuffer;
 import java.nio.charset.Charset;
 import java.security.InvalidKeyException;
 import java.security.MessageDigest;
@@ -105,7 +104,10 @@ public class TPMProcessor implements AttestationFormatProcessor {
             throw new Fido2RuntimeException("Problem with TPM attestation");
         }
 
-        byte[] hashedBuffer = getHashedBuffer(cborPublicKey.get("3").asInt(), authData.getAttestationBuffer(), clientDataHash);
+        verifyVersion2(attStmt);
+        int alg = verifyAlg(attStmt);
+
+        byte[] hashedBuffer = getHashedBuffer(alg, authData.getAttestationBuffer(), clientDataHash);
         byte[] keyBufferFromAuthData = base64Service.decode(cborPublicKey.get("-1").asText());
 
         Iterator<JsonNode> i = attStmt.get("x5c").elements();
@@ -135,7 +137,7 @@ public class TPMProcessor implements AttestationFormatProcessor {
             byte[] certInfoBuffer = base64Service.decode(certInfo);
             byte[] signatureBytes = base64Service.decode(signature.getBytes());
 
-            signatureVerifier.verifySignature(signatureBytes, certInfoBuffer, aikCertificate, authData.getKeyType());
+            signatureVerifier.verifySignature(signatureBytes, certInfoBuffer, aikCertificate, alg);
 
             byte[] pubAreaBuffer = base64Service.decode(pubArea);
             TPMT_PUBLIC tpmtPublic = TPMT_PUBLIC.fromTpm(pubAreaBuffer);
@@ -146,6 +148,9 @@ public class TPMProcessor implements AttestationFormatProcessor {
             verifyTPMSExtraData(hashedBuffer, tpmsAttest.extraData);
             verifyThatKeysAreSame(tpmtPublic, keyBufferFromAuthData);
 
+            credIdAndCounters.setAttestationType(getAttestationFormat().getFmt());
+            credIdAndCounters.setCredId(base64Service.urlEncodeToString(authData.getCredId()));
+            credIdAndCounters.setUncompressedEcPoint(base64Service.urlEncodeToString(authData.getCosePublicKey()));
         } else {
             throw new Fido2RuntimeException("Problem with TPM attestation. Unsupported");
         }
@@ -197,10 +202,9 @@ public class TPMProcessor implements AttestationFormatProcessor {
 
     private byte[] getHashedBuffer(int digestAlgorith, byte[] authenticatorDataBuffer, byte[] clientDataHashBuffer) {
         MessageDigest hashedBufferDigester = signatureVerifier.getDigest(digestAlgorith);
-        byte[] b1 = authenticatorDataBuffer;
-        byte[] b2 = clientDataHashBuffer;
-        byte[] buffer = ByteBuffer.allocate(b1.length + b2.length).put(b1).put(b2).array();
-        return hashedBufferDigester.digest(buffer);
+        hashedBufferDigester.update(authenticatorDataBuffer);
+        hashedBufferDigester.update(clientDataHashBuffer);
+        return hashedBufferDigester.digest();
     }
 
     private void verifyTPMCertificateExtenstion(X509Certificate aikCertificate, AuthData authData) {
@@ -220,5 +224,27 @@ public class TPMProcessor implements AttestationFormatProcessor {
             log.warn("Problem with AIK certificate {}", e.getMessage());
             throw new Fido2RuntimeException("Problem with TPM attestation");
         }
+    }
+
+    private void verifyVersion2(JsonNode attStmt) {
+        if (!attStmt.has("ver")) {
+            log.error("TPM does not contain the ver");
+            throw new Fido2RuntimeException("TPM does not contain the 'ver'");
+        }
+        String version = attStmt.get("ver").asText();
+        if (!version.equals("2.0")) {
+            log.error("TPM invalid version, ver 2.0 is required");
+            throw new Fido2RuntimeException("TPM invalid version, ver 2.0 is required");
+        }
+    }
+
+    private int verifyAlg(JsonNode attStmt) {
+        if (!attStmt.has("alg")) {
+            log.error("TPM does not contain the alg");
+            throw new Fido2RuntimeException("TPM does not contain the 'alg'");
+        }
+        int alg = attStmt.get("alg").asInt();
+        log.trace("TPM attStmt 'alg': {}", alg);
+        return alg;
     }
 }

--- a/jans-fido2/server/src/main/java/io/jans/fido2/service/verifier/SignatureVerifier.java
+++ b/jans-fido2/server/src/main/java/io/jans/fido2/service/verifier/SignatureVerifier.java
@@ -102,7 +102,7 @@ public class SignatureVerifier {
                 return signatureChecker;
             }
             case -65535: {
-                Signature signatureChecker = Signature.getInstance("SHA1withRSA", provider);
+                Signature signatureChecker = Signature.getInstance("SHA1withRSA");
                 return signatureChecker;
             }
 

